Output test log path (optionally contents) post-install, include post-install times

tldahlgren opened the pull request:
(UPDATED) This PR provides more post-install test information:

1. prints the test log path IF a test failure occurs;
2. prints the contents of the test log IF `-v` option is passed to `spack install`; and
3. times and prints post-install operations (e.g., post-install tests).

For example, with `-v --test=root` for a package with a (fake) test failure:
```
$ spack install -v --test=root py-libensemble@0.9.2
...
==> Installing py-libensemble-0.9.2-lr5qi7bfilbfm7ci5jvwno5kccjavc7d
==> No binary for py-libensemble-0.9.2-lr5qi7bfilbfm7ci5jvwno5kccjavc7d found: installing from source
==> Using cached archive: $spack/var/spack/cache/_source-cache/archive/e4/e46598e5696f770cbff4cb90507b52867faad5654f1b80de35405a95228c909f.tar.gz
==> No patches needed for py-libensemble
==> py-libensemble: Executing phase: 'install'
...
Successfully built libensemble
Installing collected packages: libensemble

Successfully installed libensemble-0.9.2
...
==> Testing package py-libensemble-0.9.2-lr5qi7b
==> [2023-02-24-16:08:32.534705] Running install-time tests
==> [2023-02-24-16:08:32.536319] Installing $/spack/opt/spack/linux-rhel7-broadwell/gcc-8.3.1/py-libensemble-0.9.2-lr5qi7bfilbfm7ci5jvwno5kccjavc7d/.spack/test to /var/tmp/dahlgren/spack-stage/spack-stage-py-libensemble-0.9.2-lr5qi7bfilbfm7ci5jvwno5kccjavc7d/py-libensemble-0.9.2-lr5qi7b/cache/py-libensemble
==> [2023-02-24-16:08:32.811530] RUN-TESTS: install-time tests [test]
==> [2023-02-24-16:08:32.876125] Detected the following modules: ['libensemble', 'libensemble.alloc_funcs', 'libensemble.comms', 'libensemble.executors', 'libensemble.executors.balsam_executors', 'libensemble.gen_funcs', 'libensemble.resources', 'libensemble.sim_funcs', 'libensemble.sim_funcs.branin', 'libensemble.tests', 'libensemble.tests.regression_tests', 'libensemble.tests.unit_tests', 'libensemble.tools', 'libensemble.utils']
==> [2023-02-24-16:08:32.877048] checking import of libensemble
==> [2023-02-24-16:08:32.877751] '/usr/tce/bin/python3.7' '-c' 'import libensemble'
PASSED
...
==> [2023-02-24-16:08:45.706608] test: run test_1d_sampling.py example
==> [2023-02-24-16:08:45.707351] '/usr/tce/bin/python3.7' 'test_1d_sampling.py' '--comms' 'local' '--nworkers' '2'

libEnsemble with random sampling has generated enough points
libensemble.tools.tools: ------------------ Run completed -------------------
Saving results to file: 1d_sampling_history_length=1000_evals=502_workers=2
PASSED
==> [2023-02-24-16:08:47.117321] test: fail
==> [2023-02-24-16:08:47.118023] '/usr/tce/bin/python3.7' '-c' 'raise RuntimeError("fail")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
RuntimeError: fail
FAILED: Command exited with status 1:
    '/usr/tce/bin/python3.7' '-c' 'raise RuntimeError("fail")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
RuntimeError: fail
...
  File "$spack/var/spack/repos/builtin/packages/py-libensemble/package.py", line 100, in test
    purpose="test: fail",
==> [2023-02-24-16:08:47.450596] Completed testing
==> Error: TestFailure: 1 tests failed.

Command exited with status 1:
    '/usr/tce/bin/python3.7' '-c' 'raise RuntimeError("fail")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
RuntimeError: fail

1 error found in test log:
...
See build log for details:
  $TMPDIR/spack-stage/spack-stage-py-libensemble-0.9.2-lr5qi7bfilbfm7ci5jvwno5kccjavc7d/spack-build-out.txt

See test log for details:
  $TMPDIR/spack-stage/spack-stage-py-libensemble-0.9.2-lr5qi7bfilbfm7ci5jvwno5kccjavc7d/install-time-test-log.txt

193.295u 23.579s 3:59.98 90.3%	0+0k 1816+6256io 1pf+0w
```

While running tests without `-v` for the same package results in:
```
$ spack install --test=root py-libensemble@0.9.3
...
==> Installing py-libensemble-0.9.3-5prbuneeootut5fplscnx7mv2ews3xin
==> No binary for py-libensemble-0.9.3-5prbuneeootut5fplscnx7mv2ews3xin found: installing from source
==> Using cached archive: /usr/WS1/dahlgren/releases/spack/var/spack/cache/_source-cache/archive/00/00e5a65d6891feee6a686c048d8de72097b8bff164431f163be96ec130a9c390.tar.gz
==> No patches needed for py-libensemble
==> py-libensemble: Executing phase: 'install'
==> Error: TestFailure: 1 tests failed.

Command exited with status 1:
    '/usr/tce/bin/python3.7' '-c' 'raise RuntimeError("fail")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
RuntimeError: fail
...
See build log for details:
  /var/tmp/dahlgren/spack-stage/spack-stage-py-libensemble-0.9.3-5prbuneeootut5fplscnx7mv2ews3xin/spack-build-out.txt

See test log for details:
  /var/tmp/dahlgren/spack-stage/spack-stage-py-libensemble-0.9.3-5prbuneeootut5fplscnx7mv2ews3xin/install-time-test-log.txt

199.725u 30.088s 4:00.49 95.5%	0+0k 1024+5688io 0pf+0w
``` 

Note:
```
$ spack  -v test run py-libensemble@0.9.2  # shows tests and, if a failure, log path
$ spack test run py-libensemble@0.9.3  # does not show details or the log path

TODO

- [x] add/update unit tests

tldahlgren committed:
Output test log path (optionally contents) post-install

tldahlgren committed:
Add unit tests

tldahlgren committed:
Don't output the test log if the tests weren't run

tldahlgren commented:
@adamjstewart @alalazo Resolved conflicts.  The patch coverage is reported at 100% and project coverage up +.13%.  Also re-confirmed the example output shown in the description.

adamjstewart commented on lib/spack/spack/installer.py:
```diff
@@ -280,6 +280,32 @@ def _print_installed_pkg(message):
+    """Output install test log file information.
+
+    Args:
+        pkg (spack.package.PackageBase): package of interest
```
Can you move these types from the docstring to the function signature so that mypy can test them?

tldahlgren replied:
Ouch.  I keep forgetting we changed `package` module to `package_base` too.

tldahlgren replied:
Grrr.  I keep getting `AttributeError: module 'spack' has no attribute 'package_base'`.

tldahlgren replied:
I can add `: bool` but it is refusing to work when I add `: spack.package_base.PackageBase`.

adamjstewart replied:
You might have to import package_base

tldahlgren replied:
It is already importing `spack.package_base`.

tldahlgren replied:
The module is already used for several calls (and docstrings).

adamjstewart replied:
Try adding the commit so I can see the failing tests

tldahlgren replied:
Will do.  I'm manually testing colorizing test log messages (and `Completed testing`).

tldahlgren replied:
Done

tldahlgren replied:
Interesting.  My local checks didn't add the comment that it was most likely due to a circular import.

tldahlgren replied:
Can I revert that last commit and add the typing to the refactored code in #34236?

adamjstewart replied:
If you put the base class in quotes (so it's only evaluated by the type checker and doesn't actually need to be imported) does that work?

tldahlgren replied:
I didn't realize that was an option.  🎉

adamjstewart commented on lib/spack/spack/installer.py:
```diff
@@ -1945,6 +1971,7 @@ def run(self):
             spack.hooks.post_install(self.pkg.spec)
 
         _print_timer(pre=self.pre, pkg_id=self.pkg_id, timer=self.timer)
+        _print_install_test_log(self.pkg, self.verbose)
```
I feel like this should come before the timer, and possibly be included in the time calculation. Maybe an extra step for how long the tests took to run.

tldahlgren replied:
I ended up adding a timer for handling `post-install`, which may be more than just the tests.

Just noticed I don't log `Completed` for post-install tests like I do when running stand-alone tests just for the purposes of extracting test timing from the test log (for CDash reporting).

tldahlgren replied:
Added `Completed testing` when running a `test` method post-install.

adamjstewart commented on lib/spack/spack/installer.py:
```diff
@@ -280,6 +280,32 @@ def _print_installed_pkg(message):
+            for ln in f.readlines():
+                print(ln.strip("\n"))
+
+    print("\nSee test results at:\n  {0}".format(log))
```
I don't know if we need to print this unless the tests fail. We don't print spack-build-out.txt and friends unless the install fails.

tldahlgren replied:
Fair enough.

tldahlgren replied:
My main reservation here is that it's handy to be reminded where the test outputs are in case you want to make sure all of the tests you expect to be executed (e.g., inherited, including from virtual) are actually run. 

I have also seen "passing" tests incur failures that aren't trapped/tracked by the test implementation.   I should make a list of these when I see them.

adamjstewart replied:
Likewise, it's handy to be reminded where `spack-build-out.txt` is, but I don't think we need to do that by default on every package every time. Users can use `--verbose` if they want to see which tests were executed.

tldahlgren replied:
True.  (I already changed this BTW.)

adamjstewart reviewed (commented):
Tested with py-six and works as expected! A few comments related to the output and when/where we print it. Also, the build output uses green for the ==>, but the test output uses no color. Probably simple to fix.

tldahlgren committed:
Addressed feedback for typing, post-install timing, and outputs:
- added typing for _print_install_test_log arg(s);
- added timing of post-install (to capture testing timing data);
- changed to output test log path only on failures; and
- colorized test log markers green.

Also:
- Added "Completed testing" log message for post-install tests; and
- colorized test log messages.

adamjstewart commented:
This is looking much better now. A few final things I noticed:

1. The `install-time-test-log.txt` file that gets installed is executable and prob shouldn't be
2. The font used for `==>` during testing is bold, but the font used for `==>` during build isn't
3. The test start/complete messages are printed twice:
   ```
   ==> Testing package py-six-1.15.0-6ev665z
   ==> [2023-02-23-21:12:48.859872] Running install-time tests
   ==> [2023-02-23-21:12:48.861115] RUN-TESTS: install-time tests [test]
   ==> [2023-02-23-21:12:48.911128] Completed testing
   ==> Testing package py-six-1.15.0-6ev665z
   ==> [2023-02-23-21:12:48.859872] Running install-time tests
   ==> [2023-02-23-21:12:48.861115] RUN-TESTS: install-time tests [test]
   ==> [2023-02-23-21:12:48.863071] Detected the following modules: ['six']
   ==> [2023-02-23-21:12:48.864000] checking import of six
   ==> [2023-02-23-21:12:48.864307] '/Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/python-3.10.8-vtplktw577cpgtkbjew4bjbetrxjttbu/bin/python3.10' '-c' 'import six'
   PASSED
   ==> [2023-02-23-21:12:48.911128] Completed testing
   ```

tldahlgren committed:
Fix test log perms; log mark coloring; test output duplication; unit test.

tldahlgren committed:
Only show the test log path on failure

adamjstewart commented:
With the latest commit, `spack install -v --test=root` no longer prints the test output.

tldahlgren commented:
> With the latest commit, `spack install -v --test=root` no longer prints the test output.

Can you try `spack -v install --test=root`?  I will revisit getting the option passed  the other way as well.

adamjstewart commented:
Still didn't display the tests even with:
```console
$ spack -v install -v --test=root py-six
```

tldahlgren commented:
Closing this PR since the changes originated in https://github.com/spack/spack/pull/34236, which has refactoring that causes redundant work iterating on this and the other PR.
